Handle errors when marking invoices paid

adamshire123 opened the pull request:
### What does this PR do?

Adds error handling and unit tests to a function that marks invoices 'paid' in alma. 

If a request error is encountered during an API call to mark an invoice 'paid', we log the invoice number and move on to the next invoice. The skipped invoices can be paid manually using API calls afterwards. 

### Helpful background context

We have been experiencing timeout errors when the app tries to mark invoices as paid. 

### Includes new or updated dependencies?

NO

### What are the relevant tickets?

https://mitlibraries.atlassian.net/browse/INFRA-381

### Developer

- [x] All new ENV is documented in README (or there is none)
- [x] Stakeholder approval has been confirmed (or is not needed)

### Code Reviewer

- [ ] The commit message is clear and follows our guidelines
      (not just this pull request message)
- [ ] There are appropriate tests covering any new functionality
- [ ] The documentation has been updated or is unnecessary
- [ ] The changes have been verified
- [ ] New dependencies are appropriate or there were no changes

adamshire123 committed:
Handle errors when marking invoices paid

Why these changes are being introduced:
We have been experiencing timeout errors when the app makes API calls to Alma to mark invoices paid.

How this addresses that need:
* adds try / except block to handle errors occuring during the request
* adds unit tests for error handling

Relevant ticket(s):
* https://mitlibraries.atlassian.net/browse/INFRA-381

adamshire123 commented:
I think ultimately, we want a way of not just getting notified about the skipped invoices, but also a command to pay them. At least this PR let's the app keep running if a read timeout comes up.

adamshire123 commented:
see https://github.com/MITLibraries/alma-sapinvoices/issues/120

ehanson8 commented on sapinvoices/sap.py:
```diff
@@ -630,18 +631,22 @@ def mark_invoices_paid(
+                logger.debug("Invoice '%s' marked as paid in Alma", invoice_id)
+                paid_invoice_count += 1
+            else:
+                raise ValueError
```
Do we need to test for a `ValueError` being raised as well?

adamshire123 replied:
I think so, the idea is that we may get a valid response from the Alma API, but the payment status value is not equal to "PAID", and if it isn't PAID, we need to treat it the same way as if a request exception occurred, e.g. logging an error.

ehanson8 commented on sapinvoices/sap.py:
```diff
@@ -630,18 +631,22 @@ def mark_invoices_paid(
+        except (requests.exceptions.RequestException, ValueError):
             logger.error(
                 "Something went wrong marking invoice '%s' paid in Alma, it "
                 "should be investigated manually",
```
I don't think `manually` is necessary here

adamshire123 replied:
Probably not necessary, I was just using what was there previously.

ehanson8 reviewed (commented):
A few questions

adamshire123 committed:
Check payment success or failure sooner

Move checking the response of the pay invoice API call
to the function making the API call. If the API call is successful,
but the invoice is not marked as Paid, raise a ValueError

adamshire123 committed:
fix linting error

ehanson8 reviewed (approved):
I like this approach better but it would be good to use the usual bulleted list of the significant changes (including test updates) in your commit messages. 👍
